src/serlo-editor-repo/store/focus/selectors.ts

```diff
@@ -0,0 +1,125 @@
+import { createSelector } from '@reduxjs/toolkit'
+import * as R from 'ramda'
+
+import { selectDocument } from '../documents'
+import {
+  createDeepEqualSelector,
+  createJsonStringifySelector,
+} from '../helpers'
+import { selectPlugin } from '../plugins'
+import { selectRoot } from '../root'
+import { State } from '../types'
+import { findParent } from './helpers'
+import type { FocusTreeNode } from './types'
+
+const selectSelf = (state: State) => state.focus
+
+export const selectFocused = createSelector(selectSelf, (focus) => focus)
+
+export const selectIsFocused = createSelector(
+  [selectSelf, (_state, id: string) => id],
+  (focus, id: string) => focus === id
+)
+
+export const selectFocusTree: (
+  state: State,
+  id?: string
+) => FocusTreeNode | null = createJsonStringifySelector(
+  [(state: State) => state, (_state, id?: string) => id],
+  (state, id = undefined) => {
+    const root = id ? id : selectRoot(state)
+    if (!root) return null
+    const document = selectDocument(state, root)
+    if (!document) return null
+    const plugin = selectPlugin(state, document.plugin)
+    if (!plugin) return null
+
+    const children = plugin.state
+      .getFocusableChildren(document.state)
+      .map((child) => {
+        const subtree = selectFocusTree(state, child.id)
+        return subtree || child
+      })
+
+    return {
+      id: root,
+      children,
+    }
+  }
+)
+
+export const selectParent = createSelector(
+  [(state: State) => state, (_state, id: string) => id],
```

Comment:
For sure, I was thinking about it for some time. In general, the `selectFocusTree` is used all, not only for focusing feature. And it has to select state from documents and plugins, which makes me wonder if we should re-think the whole state structure.

In any case, if you think of a way to rewrite this selector without changing the state structure, please feel free to add the change, whenever. Otherwise, we can think about the problem a bit deeper in the future.